### DISCUSSION TITLE
Why does this GraphQL search query return empty nodes?

### DISCUSSION THREAD

**Discussion by test21324** [community/community#78663] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I'm using the GraphQL API specifically the search object to query users in GitHub.
Why does this return empty nodes? I'm getting less than 1000 items.

![image](https://github.com/community/community/assets/151022857/5e7cd4fe-0bf8-48c5-94ce-e43947de4173)

**Comment by test21324** [comment]
Any help on this, or any other place I can ask for support?

**Reply by LangLangBart** [reply]
> Why does this return empty nodes?

> [!TIP]
> > USER: Returns results matching users and organizations on GitHub.


```graphql
{
  __type(name: "SearchType") {
    enumValues {
      name
      description
    }
  }
}
```

```json
{
  "data": {
    "__type": {
      "enumValues": [
        {
          "name": "ISSUE",
          "description": "Returns results matching issues in repositories."
        },
        {
          "name": "REPOSITORY",
          "description": "Returns results matching repositories."
        },
        {
          "name": "USER",
          "description": "Returns results matching users and organizations on GitHub."
        },
        {
          "name": "DISCUSSION",
          "description": "Returns matching discussions in repositories."
        }
      ]
    }
  }
}
```

- use of `... on Organization`

```graphql
{
  search(first: 100, type: USER, query: "created:>=2023-12-12T06") {
    userCount
    nodes {
      ... on Organization {
        id
        __typename
      }
      ... on User {
        id
        __typename
      }
    }
  }
}
```

```json
{
  "data": {
    "search": {
      "userCount": 15337,
      "nodes": [
        {
          "id": "U_kgDOCSeJkQ",
          "__typename": "User"
        },
        {
          "id": "U_kgDOCSeLXg",
          "__typename": "User"
        },
        {
          "id": "O_kgDOCSekKQ",
          "__typename": "Organization"
        },
        {
          "id": "U_kgDOCSea6Q",
          "__typename": "User"
        },
        {
          "id": "U_kgDOCSfWCQ",
          "__typename": "User"
        }
      ]
    }
  }
}
```

---

> I'm getting less than 1000 items.

> [!TIP]
> > GitHub's GraphQL API limits the number of items that you can fetch in a single request in order to protect against excessive or abusive requests to GitHub's servers. When you use the GraphQL API, you must supply a first or last argument on any connection. The value of these arguments must be between 1 and 100.
> Source: [Using pagination in the GraphQL API - GitHub Docs ](https://docs.github.com/en/graphql/guides/using-pagination-in-the-graphql-api)

- For this query, you need to keep the query running each time you get `true` from the
  `hasNextPage`. Then, run the subsequent query with the `endCursor` as input.

```graphql
query ($endCursor: String) {
  search(first: 100, after: $endCursor, type: USER, query: "created:>=2023-12-12T10:15") {
    userCount
    pageInfo {
      hasNextPage
      endCursor
    }
    nodes {
      ... on Organization {
        id
        __typename
      }
      ... on User {
        id
        __typename
      }
    }
  }
}
```

- The `gh` command simplifies pagination by offering a `--paginate` flag. This flag allows the command to continue running until there are no more pages or until the rate limit is reached.

```bash
gh api graphql --paginate --raw-field query='query ($endCursor: String) {
  search(first: 100, after: $endCursor, type: USER, query: "created:>=2023-12-12T10:15") {
    userCount
    pageInfo {
      hasNextPage
      endCursor
    }
    nodes {
      ... on Organization {
        id
        __typename
      }
      ... on User {
        id
        __typename
      }
    }
  }
}'
```

---

Informative talk about GraphQL by Rea Loretta (15/Dec/19) - [Advanced patterns for GitHub's GraphQL API](https://www.youtube.com/watch?v=i5pIszu9MeM)

**Comment by test21324** [comment, marked as answer]
I solved this issue, the SearchType type:USER returns _users and organizations_.
![image](https://github.com/community/community/assets/151022857/de9805c9-992e-42db-a044-5571a70d8a0b)
